### DISCUSSION TITLE
CSS is not loaded in rewrites(External URL)

### DISCUSSION THREAD

**Discussion by iamsriranga** [vercel/next.js#20650] [Help]
I want to migrate my existing react project to nextjs, meanwhile, I came across this rewrites in nextjs so I implemented their [with-rewrites](https://github.com/vercel/next.js/tree/canary/examples/rewrites) by changing one of the routes to "reactnative.dev/help" but the problem is HTML is being loaded but not CSS. Any suggestions/thoughts.

Here's the link [next-rewrites-ex](https://next-rewrites-bqwau6b2y.vercel.app/help)

Here's the [repo](github.com/saideepesh000/next-rewrites)

**Comment by lachlanjc** [comment]
If I'm understanding correctly, this is intentional behavior—rewrites serve the complete page at the destination URL with no interference from the Next app itself. It's a server-level response, not an app-level one, & so doesn't bundle any of the CSS of the Next application.

If the CSS loaded by the remote page isn't working correctly, check if it's a relative path (/something.css), & if so, add a rewrite for that path to the external domain.

**Reply by heli-solutelabs** [reply]
Hi @lachlanjc what does the above solution mean by relative path(/something.css).
I am stuck with the same issue.
Can u help me elaborating a little more about the solution u proposed???

**Reply by lachlanjc** [reply]
In the HTML file, there's likely links to CSS & JS files with relative URLs (starting with a slash), which will be loaded from your domain via the rewrite, but they won't exist, since you only added the rewrite for the HTML file. You'll need to make those assets resolve via additional rewrites to their origin server, since they don't exist on your domain. In the CSS or JS files there may be further relative links; check for `@import` or `url()` that starts with a slash in the CSS.

**Comment by ghost** [comment]
facing same issue,
when the main app loads the external redirects, it doesn't load the static folder of the redirected app (**for ex:** _next/static/).
It is trying to access static assets from the main app's static folder which doesn't contain the required assets for the redirected app.

**Comment by Emily-Thomas** [comment]
Having the same issue, css, images, and JS refuse to be loaded. only thing that is loaded is in the initial HTML file

**Comment by subhasishdas159** [comment]
I'm facing the same issue with rewrites. Please suggest a how to fix.

**Comment by wiiiimm** [comment]
Can you please try to take a look at this example: https://github.com/vercel/next.js/tree/canary/examples/with-zones ?
Based on this example, I have created another example at https://40049-nextjs-with-zones.vercel-support.app using the repo at https://github.com/vercel-support/40049-nextjs-with-zones.

**Comment by DevSachinTiwari** [comment]
I'm facing the same issue with rewrites. Please suggest a how to fix.

**Comment by normdoow** [comment]
I'm seeing this issue as well. But it only happens when you are trying to rewrite to another site that is also a next.js site.
I believe it happens because Next trys to load the site's .js and css files (/_next/static files) from your own next.js app instead of the one you are rewritng to.

<img width="266" alt="image" src="https://user-images.githubusercontent.com/21371266/202793396-0dd0c512-1e9b-4586-84ed-0d9c6319ec09.png">

**Reply by jespanag** [reply]
@normdoow do you find any solution ?  :)

**Reply by jespanag** [reply]
Ready 👇🏻

**Comment by jespanag** [comment]
### I found a tricky solution:

**‼️ Only works if you have access to both domains**

**Website1.com** that rewrites to **website2.com** use the normal configuration:

In `next.config.js`
```js
async rewrites() {
    return {
      beforeFiles: [
        {
          source: "/example",
          destination: "http://website2.com/example",
        },
      ],
    }
```

Now, the problem is that, because `website2.com` is also executed with *nextjs*, all scripts are loaded from a dynamic path `/nextjs/....`, when `website1.com` tries to load these dynamic routes, it does from itself (Obviously it doesn't exist), so we need to convert all dynamic routes in `website2.com` to absolute:

**In website2.com**:

In `next.config.js` add: `assetPrefix` with the matching domain value (`website2.com`)
```js
const nextConfig = {
  ...
  assetPrefix: 'http://website2.com',
  ...
}
```

Now your `website2.com` will load the nextjs scripts from an absolute path, and `website1.com` will get it correctly.

I hope it works for you 😄

**Reply by upyo-ankit** [reply]
@jespanag thank you so much... this solution help me a... I was finding solution on this problem since 3 days....
now only problem with me is

```
beforeFiles : [
        {
          source: '/services',
          destination: `${SERVICES_URL}/services`
        },
        {
          source: '/services/:path*',
          destination: `${SERVICES_URL}/services/:path*`
        }
      ],
```
services page working fine but from that page if I click on any link /services/path redirecting back to main domain "http://localhost:3000/:path"  and I am getting 404 Page not found error

subdomain path "http://localhost:4000/services/:path" is working fine

Main domain:  http://localhost:3000      
Subdomain : http://localhost:4000      
      
      Any idea what might be the problem ?
      
      Thanks again for the solution

**Reply by jespanag** [reply]
@upyo-ankit idk the solution but, first try to remove the `beforeFiles`.

1. Without `beforeFiles`:

```js
 async rewrites() {
    return [
        {
          source: '/services',
          destination: `${SERVICES_URL}/services`
        },
        {
          source: '/services/:path*',
          destination: `${SERVICES_URL}/services/:path*`
        }
    ];
  },
...
```

2. Check that in you `redirects()` you don't have any redirection from `/services/:path*`

3. Like a very strange option... use regex instead of `:path*` (i use this to rewrite from wordpress)

```js
 async rewrites() {
    return [
        {
          source: '/services',
          destination: `${SERVICES_URL}/services`
        },
        {
          source: '/services/(.*?)([.]\\w{2,5})(\\?[^/]+)?',
          destination: `${SERVICES_URL}/services/$1$2$3*`
        },
        {
          source: '/services/([\\S\\s]+)',
          destination: `${SERVICES_URL}/services/$1/`
        }
    ];
  },
...
```

Try if any of this is usefull

**Reply by AryanJ-NYC** [reply]
> ### I found a tricky solution:
> **‼️ Only works if you have access to both domains**
> 
> **Website1.com** that rewrites to **website2.com** use the normal configuration:
> 
> In `next.config.js`
> 
> ```js
> async rewrites() {
>     return {
>       beforeFiles: [
>         {
>           source: "/example",
>           destination: "http://website2.com/example",
>         },
>       ],
>     }
> ```
> 
> Now, the problem is that, because `website2.com` is also executed with _nextjs_, all scripts are loaded from a dynamic path `/nextjs/....`, when `website1.com` tries to load these dynamic routes, it does from itself (Obviously it doesn't exist), so we need to convert all dynamic routes in `website2.com` to absolute:
> 
> **In website2.com**:
> 
> In `next.config.js` add: `assetPrefix` with the matching domain value (`website2.com`)
> 
> ```js
> const nextConfig = {
>   ...
>   assetPrefix: 'http://website2.com',
>   ...
> }
> ```
> 
> Now your `website2.com` will load the nextjs scripts from an absolute path, and `website1.com` will get it correctly.
> 
> I hope it works for you 😄

Your hack worked! Thank you! It saved me a ton of time!

**Reply by olafkotur** [reply]
👏🏽 You. Deserve. A. Medal. 👏🏽

**Reply by nathanmrtns** [reply]
I tried here, but it doesn't work. I guess it is because my destination application is not nextjs. Any idea?

**Comment by huyjs9** [comment]
### The issues I face in my context:
Root domain: https://example.com
Subdomain: https://docs.example.com
I want to config that if you access https://example.com/docs or https://www.example.com/docs, it will proxy to https://docs.example.com and keep the URL.
- In `next.config.js`:
```js
module.exports = {
  trailingSlash: true,
  async rewrites() {
    return [
      {
        source: '/docs/:path*/',
        destination: 'https://docs.example.com/:path*/',
      },
    ]
  },
}
```
It cannot load the CSS.

### My solution here:
- In `next.config.js`:
```js
module.exports = {
  trailingSlash: true,
}
```

- In `middleware.js`:
```js
import { NextResponse } from 'next/server';

export default async function middleware(req) {
  const { pathname } = req.nextUrl
  if (pathname.startsWith('/docs')) {
    return NextResponse.rewrite(new URL(pathname, "https://docs.example.com"))
  }
  return NextResponse.next()
}
```
I hope it works for you.

**Comment by GabrielSoares-Dev** [comment]
How to fix this when use next host and vitejs other application?

**Comment by zohairhadi** [comment]
My issue was resolved by doing 2 things
1. removing all relative imports and replacing them with absolute imports in the target repo code.

example
`import 'globals.css'` -> `import "@/app/globals.css"`

2. Adding a path rewrite to the next.config to the origin server like @lachlanjc said here: https://github.com/vercel/next.js/discussions/20650#discussioncomment-4287353
